astropy/modeling/parameters.py

```diff
@@ -416,6 +416,16 @@ def internal_unit(self, internal_unit):
         """
         self._internal_unit = internal_unit
 
+    @property
+    def input_unit(self):
+        """Unit for the input value."""
+        if self.internal_unit is not None:
+            return self.internal_unit
+        elif self.unit is not None:
+            return self.unit
+        else:
+            return None
```

Comment:
This is needed when the `internal_unit` and the `unit` disagree. Computation happens on the internal values, if they are different from the main value otherwise the internal value is not set. This is so that the input unit checks can correctly "guess" the unit for the input value, which in this case should match the internal values.

The model changes in:

- `functional_models.py`
- `physical_models.py`
- `polynomial.py`
- `powerlaws.py`
- `projections.py`

Exist to enable unit support when one uses `getter/setter` `ufuncs` to alter the parameter space of the value being passed into the parameter, such as the case in #14511. These models will need to guess on the internal unit values in these cases via the `input_unit` property above.